### PR TITLE
Detect state hash inconsistency before commit

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
Note: this builds on top of https://github.com/radixdlt/babylon-node/pull/312

Here is an example usage of state hash (to detect engine non-det bugs).

### CODE REVIEW

**Comment by jakrawcz-rdx** [comment]
@talekhinezh please let me know if you prefer to `panic!` (instead of `warn!`) on engine non-determinism.

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommitRequest.java [hunk 1]
```diff
@@ -77,6 +78,7 @@
 public record CommitRequest(
     List<RawLedgerTransaction> transactions,
     UInt64 stateVersion,
+    HashCode stateHash,
```
Comment (dhedey) [inline core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommitRequest.java:81]: I might be tempted to call it `expectedResultantStateHash` or something, but that's also me loving long descriptive names.

Happy to keep with `stateHash` too :D

File: core-rust/state-manager/src/types.rs [hunk 2]
```diff
@@ -480,10 +480,11 @@ pub enum CommitError {
     MissingEpochProof,
 }
 
-#[derive(Debug, Decode, Encode, Categorize)]
+#[derive(Debug, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
```
Comment (dhedey) [inline core-rust/state-manager/src/types.rs:483]: We should be able to just use `Decode`/`Encode`/`Categorize` here - (and maybe `StateHash` if necessary?).

ScryptoX is only needed for types which require the Scrypto SBOR custom value model (ie things referencing runtime engine types).

**Review by dhedey — APPROVED** [review]
Looks good!

File: core-rust/state-manager/src/types.rs (hunk 2 above)
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/types.rs]: ahh, indeed ✅

**Pull request merged**